### ISSUE TITLE
interface of benchmark to algorithms

### ISSUE THREAD

**Issue by martinjaggi** [mlbench/mlbench-old#58]
we've discussed this with @Panaetius  today, and in the medium-term we'd like to clean up the interface and make it much simpler, both for the closed and open divisions.

the goal is that one can basically bring existing distributed training algorithms, and just add two lines of python. one is doing `import mlbench` , and then at any appropriate place inside some training loop, the algorithm can call `mlbench.eval(model)`.
a big advantage of this would also be that it would become easier to reuse the provided reference implementations for other purposes.

there are some decisions to make still, including how to split the configs. for example the eval needs to know where the test dataset is, while the train algo needs to know the train dataset only.

suggestions welcome!

**Comment by LiamHe** [comment]
As a continuation of our discussion, the structure of reference could be 

<img width="221" alt="screen shot 2018-09-12 at 15 21 23" src="https://user-images.githubusercontent.com/13152755/45427587-87d8fc80-b69f-11e8-8c46-acbbd9d21123.png">

* The `parsers` are used to decide which arguments are enabled/disabled for open/close division.
* `parsers` also determines the actual behavior of `mlbench.eval`.

**Comment by martinjaggi** [comment]
perfect! so top level discrimination will be the benchmark task name or id

tiny detail: 'optim' can be merged into 'algorithms', right?

**Comment by LiamHe** [comment]
yes. I will put `lr` and `optimizer` to algorithm. 

In pytorch call the main loop [controlflow](https://pytorch.org/tutorials/beginner/examples_nn/dynamic_net.html); in mlperf they call it [run_loop](https://github.com/mlperf/training/blob/master/image_classification/tensorflow/official/resnet/resnet_run_loop.py).

<img width="211" alt="screen shot 2018-09-14 at 10 04 30" src="https://user-images.githubusercontent.com/13152755/45537947-f041ee00-b805-11e8-88fb-e730a7507b63.png">

**Comment by Panaetius** [comment]
Not sure about task at top level. I prefer a semantic hierarchy that describes what a thing is, not what it can be used for, since we don't know what a user might use something for. Our experiments are just our experiments, whereas the `mlbench` package should be a utility to use the whole mlbench framework. It shouldn’t know what experiments we have, instead being something that allows us to implement those experiments more easily. Therefore the modules and submodules should reflect what each part can be used for in general, not specifically for our experiments.

Also, I don't think we actually need to have separate namespaces/modules for tensorflow and pytorch. You can't use `pytorch.datasets.cifar10` with `tensorflow.models.resnet18`, that wouldn't make sense. Since we mentioned Keras on Wednesday, they don't have separate implementations for everything, they have three backend implementations, you set the backend once, and everything else just is the same and uses the the wrapped functions. pytorch distributed does the same for communications backend, with `init_process_group` having to be called with a `backend` parameter. We don't want to reinvent keras, but having different framework implementations in the same file and the user just having to specify once which backend to use makes a lot of sense. The backend in our case is ML framework (pytorch, tf, …) and communications framework (MPI, NCCL, …).

Small aside regarding communication backends. For OpenMPI we need to call `mpirun` with the script as command on the rank-0-worker and openmpi will take care of it. For other backends, we will probably have to start the script on each worker individually or use something similar to https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py on rank 0.

Regarding parsers, especially the distinction between open&closed division. We have to think about what is actually set by the dashboard when running an experiment vs. what is hardcoded/only relevant to an experiment. Parsers are based on the current parsers in https://github.com/mlbench/mlbench/blob/develop/mlbench/refimpls/pytorch/utils/parser.py as far as I understood. 

But for open division those are not needed at all, since a user will just have his own `main.py` in his docker and pick&choose what to use from the `mlbench` package. And the dashboard will just run `main.py` (directly or with openmpi) with few flags (number of workers, rank and ip adress of master or other nodes), plus a freeform text field that allows the user to specify additional parameters. But the format is completely up to the user. To the parser is trivial in this case plus a users (preexisting) script will probably come with its own parser already anyways.

For closed division, each task and subtask has its own config predefined, we shouldn’t allow to set all parameters on the dashboard. Since each task and subtask will be for a specific experiment, with specific settings (backend, communications framework, hyper-parameters etc. should be fixed). The machine type/number of cores is fixed, too. We don’t want hundreds of different benchmarks or all possible combinations, so for each task we pick one or two common reference datasets, one or two standard models, plus backends. The rest is fixed.

I propose the dashboard will have to have a config of all experiments that the user can chose from, something along the lines of

```yaml
- image_recognition:
  -cifar10:
    -pytorchCpuOpenmpiResnet18:
      name: pyTorch Resnet18 with OpenMPI (CPU only)
      command: /usr/bin/openmpi
      script: /app/run_loop_resnet18_cifar10_cpu.py
      image: mlbench/mlbench-pytorch-openmpi
      imageVersion: 0.1.0
      machine:
        cores: 4
        gpus: 0
        memory: 2gb 
    -pytorchCpuNCCLResnet18:
      name: pyTorch Resnet18 with NCCL (CPU only)
      command: /usr/bin/python
      script: /app/run_loop_resnet18_cifar10_cpu.py
      image: mlbench/mlbench-pytorch-nccl
      imageVersion: 0.1.0
      machine:
        cores: 4
        gpus: 0
        memory: 2gb
  -imagenet:
    -pytorchCpuOpenmpiResnet18:
      name: pyTorch Resnet18 with OpenMPI (GPU)
      command: /usr/bin/openmpi
      script: /app/run_loop_imagenet_cifar10_gpu.py
      image: mlbench/mlbench-pytorch-openmpi
      imageVersion: 0.1.0
      machine:
        cores: 4
        gpus: 1
        memory: 2gb
```

And the `run_loop_imagenet_cifar10_cpu.py` and `run_loop_resnet18_cifar10_cpu.py` have the respective hyperparameters etc. The scripts be too long and just use the relevant `mlbench` objects and execute them. The user can just select “Imagenet pyTorch Resnet18 with OpenMPI (GPU)” in the dashboard and how many nodes and start the run. The dashboard creates the StatefulSet with the correct docker image and resource requests/limits and executes the specified command. The dashboard **does not** set hyper parameters or anything else like this. This is up to each predefined experiment and is set in its respective `run_loop*.py`. The `run_loop*.py` is only called on worker 0 and takes as parameter its rank and all participating nodes (implementation depends on OpenMPI/Not-OpenMPI, Not-OpenMPI, `run_loop*.py` on worker 0 starts the script on all other workers similar to pytorch `launch.py`). This means the parser in each of those entrypoint scripts only needs to read 2-3 options from the commandline, so I don’t think we need a complex parser. Plus one parse is probably enough, since all arguments are generic enough to be applicable to all experiments. So I’d situate it in the `utils` namespace (Doesn’t make sense to have a separate namespace for a single object). This also makes it trivial for a user to run the experiment himself without dashboard or kubernetes. The images themselves should be in a separate image repository on github and mainly contain the dockerfiles +  relevant `run_loop*.py`, i.e. code specific to the experiments. 

Evaluation is complex and probably needs a specification of its own. It’s not really task specific, rather some of it is universal (TopKAccuracy calculation can be used for many things, BLEU score is applicable to many tasks etc.), some of it depends on the dataset (Test Data) and some of it depends on the model (getting predictions, loading checkpoints of we use checkpoints). The calculations themselves like TopKAccuracy can to be in their own namespace and do not depend on the framework. Getting test set data is somewhere in the dataloader and and model dependent things really depend on the model and framework (this will be especially tricky to deal with if we want to evaluate checkpoints of user models in the open division, where we have no clue what the code looks like). So the full evaluation is spread across multiple classes/functions. Having some main function where you specify the dataset, the model (i.e. a function that takes data as input and returns predictions as output) plus a list of metrics to evaluate make sense to me. Especially since we don’t want to run predictions multiple times if we calculate multiple metrics. We could have a similar function that takes a list of checkpoint definitions (Path, epoch, time etc.), a dataloader, a function that takes a checkpoint path and data as input and returns predictions plus a list of metrics to calculate, to get performance at the end of an experiment. Though this should be run on the worker as well at the end, not in the dashboard. Evaluation metrics as well as evaluation coordination can go in the evaluation namespace.

I  don't like the `algorithms` name, it’s too generic. An algorithm can be anything. It doesn't tell a user what's inside. Could be quicksort, could be a instructions for a burger making robot. I think we need a name that makes it clear to a user what to expect inside. Or if we just want some sort of catch-all for smaller things, we already have that in `utils` (which implies small, reusable things), so `algorithms` seems redundant. If there’s many different optimizer or learning rate scheduler implementations, they should reside in their own module, if not, we can just put them into a submodule of `utils`.


That would leave us with:

```
mlbench
├── api_client (formerly logging)
├── datasets
│   ├── cifar10
│   └── imagenet
├── evaluation
│   ├── metrics
│   └── coordination
├── models
│   ├── resnet18
│   └── alexnet
└── utils
    ├── argument_parser
    ├── lr
    └── optim
```



And for repositories I suggest:

- mlbench-helm: Contains helm charts (since the official helm repository requires a separate repository for charts anyways). This also tracks the main version of the whole project (in the Chart.yaml)
- mlbench-core: The python package, structured as a plain python package repository with `setup.py` etc. Get directly deployed to Pypi. This will make the repository conform to common python package layout and make it easy for users familiar with python packages to orient themselves. 
- mlbench-dashboard: Since it’s a Django application and has no direct dependencies on any of the other stuff. Contains Code + Docker
- mlbench-benchmarks: Contains docker files + experiment specific python files ( e.g. `run_loop*.py`) for the experiments.

What do you think?

**Comment by martinjaggi** [comment]
note that when you said `experiment`, each of these is in its essence a tuple of (`benchmark task`, `implementation`). 
the `benchmark task`s are most crucial and need to be most clearly visible, as they will be written in stone by us. for each task, we'll provide two more things: a evaluation (fixed by us, as it's part of the task definition), and a list of one or more implementations.
of the implementations of a task, one will be marked as our official reference implementation. others might just be there, and more might even be user contributed or black box. (therefore separation of tasks and implementation is very crucial). for example it can well be that the eval will use pytorch, while one of the implementations will be TF. (the interface of the two should allow this since it is very simple, only uses an inference function callback, taking a datapoint and giving a prediction)

what's meant by `algorithm` is training algorithm. this on one hand, and the underlying systems (hardware and frameworks) are the elements we want to benchmark.

so in terms of code structure we'd have a bit of a mix between what you and lie proposed, but with the benchmarks tasks on top level. utils of course can be shared if possible, as you mention.

the repository structure you mention sounds good too.

**Comment by Panaetius** [comment]
So if I understand you correctly, you want to have all code in the mlbench package, have the package installed in the docker and then start it with something like `/usr/bin/openmpi python -m mlbench.image_classification.pytorch.run_loop.py`?

I wouldn't want to integrate every user submitted implementation in the package, having to make a new release of the package each time and having it sort of grow organically over time. That sounds like quite a mess.

On the other hand, I think our reference implementations should mirror how a user would do his own implementation as closely as possible, for the principle of dogfooding.

So we should do our reference implementations under the viewpoint "Suppose I'm a user of MLBench, what would be the easiest way for me to run my own code in the framework". And python packages are akin to software libraries in general, with the main purpose being to group functionality so it can be easily reused. Which is also what I'd want for third party users, to give them some easy tools they can add to their code to make things work.

So I'd put the actual reference implementations in the mlbench-benchmarks repo and only keep functionality in `mlbench` that makes sense to be shared across implementations, not the implementations themselves. We might even not include model code in `mlbench`, since those are already quite implementation specific. Plus `mlbench` is where code for communication with the dashboard lives along with evaluation functions.

And the implementations in the mlbench-benchmarks repo could look something like this:

```
mlbench-benchmarks
+-- image_reconition
¦   +-- cifar10
¦   ¦   +-- task_definition.txt
¦   ¦   +-- resnet18_pytorch
¦   ¦   ¦   +-- Dockerfile
¦   ¦   ¦   +-- run_loop_openmpi_cpu.py
¦   ¦   ¦   +-- run_loop_openmpi_gpu.py
¦   ¦   ¦   +-- run_loop_nccl_cpu.py
¦   ¦   ¦   +-- run_loop_nccl_gpu.py
¦   ¦   +-- resnet18_tensorflow
¦   ¦   ¦   +-- Dockerfile
¦   ¦   ¦   +-- run_loop_openmpi_cpu.py
¦   ¦   ¦   +-- run_loop_openmpi_gpu.py
¦   ¦   ¦   +-- run_loop_nccl_cpu.py
¦   ¦   ¦   +-- run_loop_nccl_gpu.py
¦   +-- imagenet
¦       +-- task_definition.txt
¦       +-- resnet18_pytorch
¦       ¦   +-- Dockerfile
¦       ¦   +-- run_loop_openmpi_cpu.py
¦       ¦   +-- run_loop_openmpi_gpu.py
¦       ¦   +-- run_loop_nccl_cpu.py
¦       ¦   +-- run_loop_nccl_gpu.py
¦       +-- resnet18_tensorflow
¦           +-- Dockerfile
¦           +-- run_loop_openmpi_cpu.py
¦           +-- run_loop_openmpi_gpu.py
¦           +-- run_loop_nccl_cpu.py
¦           +-- run_loop_nccl_gpu.py
+-- sentiment_analysis
....
```

Possibly with .txt files alongside each implementation to describe the implementation. And additional, implementation-specific code files. But all of them use `mlbench` where appropriate for common functionality. Each of those `run_loop_*.py` files corresponds with a ` (benchmark task, implementation)` as described by you. And each of those files should be similar to what a users' `main.py` could look like with mlbench added, so they serve as nice implementation examples at the same time. 

But maybe I misunderstood you.

Could you elaborate when and why having eval in tf and training with pytorch is needed? Because we plan on having different Dockerfiles for different platforms, so at least for now there would be no Docker that could run both, so with the way we planned the implementation so far, this would not be possible.

**Comment by martinjaggi** [comment]
exactly, this would be great! i think you perfectly formulated it with the user perspective, 'bring my code to the framework', or same if one wants to take it out.

and sure i agree that user provided submissions we'll not package. but as you said with dog fooding, we want to show them it's easy to incorporate their newest fancy ML training code. so of our main principles should be that the users code should look almost the same, with or without connecting to the benchmark.

i don't want to be pushy of course and if others prefer a different architecture might be easier...

this way also the docker stuff and number of images looks quite reasonable, right?

BTW, does anybody feel strongly about `run_loop`? wouldn't `run_training` make more sense?

**Comment by martinjaggi** [comment]
### some more thoughts on providing a clean interface between (user or any) implementations and the evaluation

the 'eval in pytorch and algorithm in TF' thing is not entirely trivial indeed. and i'm fine if as you say the dockers are now implementation specific. now let's say a new user provides for example a keras implementation. the question how it will interface with the evaluation (say train error).
do we trust them that they compute the train error themselves, and correctly? and that they actually use the NN they claim to use (prescribed in the benchmark task)? this is bad and very error prone.
the trust model should be that we trust the mlbench part (which ideally contains eval functionality), but we don't trust anything in the implementation code, ideally. the tricky bit is where to put the model. (having both pytorch and TF in the same namespace should not be an issue as you mention)

Here are 3 options how to define the interface to the evaluation

1. (unsafe) **all by user** the user writes their own eval and model and we trust the whole package (maybe they use some of our utils, or not)

2. (middle ground) **callback per datapoint**. the implementation provides a prediction function. our eval calls this for say the test set points, and computes the metrics from that. at least then the metrics are correct, but only under the assumption that the users predict function is doing what it should (correctly implementing the model).

3. (safest) **model checkpoints**. the training code does whatever it wants, but from time to time it will write the model weights to disk, in a standard, framework-independent format. that is [ONNX](https://onnx.ai/) currently. then our mlbench eval code (say pytorch) later picks up that file and computes the metrics. what's nice about this is that it will also check if the NN architecture is what it claims to be

**Comment by LiamHe** [comment]
PyTorch and TensorFlow are quite different in their designs. There are not much we can share/reuse between their functions except for `parsers`, `logging`, `evaluations` and some other utility functions.  At a very high level abstraction, we can use **same function names and signatures** for PyTorch and Tensorflow implementation. Thus we can choose the backends in `mlbench.__init__.py` and call these functions with same parameter.  But at a lower level, maybe it is better to keep their own style. 

The `run_loop/run_training/controlflow` typically have very similar structures.  One only needs to make a few changes (like converting model, criterion, data to CUDA) to transform `run_loop_openmpi_cpu.py` into `run_loop_openmpi_gpu.py`. For closed divisions we should keep it away from users and allow users to replace some modules and hyperparameters. For open divisions, the `run_loop` can be customized as well.

**Comment by LiamHe** [comment]
Since users need to submit both source code and results, it seems enough for us to trust their results if : 1) there is no cheating in the submitted code 2) the submitted code can reproduce the claimed results 3) use key modules of mlbench. Do we need anything else? If it works, then we can trust implementations with more freedom.

**Comment by LiamHe** [comment]
The `model checkpoints` interface also works well with asynchronized training.

**Comment by Panaetius** [comment]
@martinjaggi ONNX would work for simple networks. But as soon as someone implements a custom layer or one framework has a feature that's not supported by our eval framework, it would break apart. We might offer it as an option, but if we enforce it as an option, we would likely exclude many models and increase the barrier of entry. But if we don't enforce it, it doesn't give us additional security. Currently there's 16 open issues for missing ops and 14 open issues for missing specs in ONNX. And since researchers are more likely to want to benchmark new, innovative models, a higher percentage of our users will use new, custom code, making it a bigger problem.
It's a further layer where things can break, there might be issues where ONNX itself influences performance in a bad way (bad implementation in ONNX or there's a bug in tf that's not in the original pytorch etc.) and it increases the effort needed by users (or us) to add their code. I'm against this option.

The middle ground sounds fine and is what I originally envisioned.

> For closed divisions we should keep it away from users and allow users to replace some modules and hyperparameters

I think in closed division, users cannot change anything at all. Everything is a given, fixed in our docker images.

> Since users need to submit both source code and results, it seems enough for us to trust their results[...]

Agreed, each user submission should be self-contained. I.e. a Dockerfile with source code that can be executed in mlbench, and sends evaluation back to reproduce the results. We can review that to make sure there's no trickery/cheating. 

Maybe we can also allow just metrics files (JSON) without source code etc. (since some big players don't like to publish their code, for copyright/trade secret reasons), but then in the open divison in the leaderboard, we can have a green ✔ or something that says "verified" if they release the source. Then we're inclusive but still make it easy to distinguish and give people an incentive to publish their code, without forcing them to. And if they provide the source, we can see if they called our eval function correctly or if there's something fishy, plus everyone else can see it, too, if we publish it in an open-division repo.

In the end, we don't have to trust the users, our users have to trust the other users. We have to be transparent about our approach, make sure we're impartial and give users the tools to make their own judgement. So we don't have to 100% fully monitor every aspect of everything, as long as we make it really clear to users what we do check and what is enforced and give them the tools to make up their own minds.

**Comment by martinjaggi** [comment]
agreed, perfect!
